<div style="text-align: center;">
  <img src="https://github.com/Hack-io-Data/Imagenes/blob/main/01-LogosHackio/logo_naranja@4x.png?raw=true" alt="esquema" />
</div>

# Laboratorio Agrupamiento de Datos

En este laboratorio usaremos el DataFrame creado en la lección anterior resultado de la unión de los conjuntos de datos compartidos contigo. 

**Instrucciones:**

1. Lee cuidadosamente el enunciado de cada ejercicio.

2. Implementa la solución en la celda de código proporcionada.

3. Devuelve los resultados siempre en formato de DataFrame

### 0. Carga los datos

In [4]:
import pandas as pd 
import numpy as np

df = pd.read_csv(r"datos_entrada\netflix.csv", index_col=0)

df.head(5)

title        Genre         Premiere  Runtime  IMDB Score  \
0   Dick Johnson Is Dead  Documentary  October 2, 2020     90.0         7.5   
1          Blood & Water          NaN              NaN      NaN         NaN   
2              Ganglands          NaN              NaN      NaN         NaN   
3  Jailbirds New Orleans          NaN              NaN      NaN         NaN   
4           Kota Factory          NaN              NaN      NaN         NaN   

  Language  originales show_id     type         director  \
0  English        True      s1    Movie  Kirsten Johnson   
1      NaN       False      s2  TV Show              NaN   
2      NaN       False      s3  TV Show  Julien Leclercq   
3      NaN       False      s4  TV Show              NaN   
4      NaN       False      s5  TV Show              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA        NaN   
3  September 24, 2021          2021  TV-MA        NaN   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

#### Ejercicio 1: ¿Cuántos títulos hay por cada tipo de contenido ('Movie' o 'TV Show')?

In [23]:
df_titulos = df.groupby("type")["title"].count().reset_index(name='title_count')

print(df_titulos)

df_titulos['title_%'] = ((df_titulos['title_count'] / df_titulos['title_count'].sum()) * 100).round(2)

df_titulos
 



#df_formaciones = df.groupby("Gender")["TrainingTimesLastYear"].sum().reset_index()
#df_formaciones["%"] = (df_formaciones["TrainingTimesLastYear"] / df_formaciones["TrainingTimesLastYear"].sum() * 100).round(2)
#df_formaciones

      type  title_count
0    Movie         6131
1  TV Show         2676


type  title_count  title_%
0    Movie         6131    69.62
1  TV Show         2676    30.38

Hay 6131 titulos del tipo "Movie"
Hay 2676 titulos del tipo TV Show

#### Ejercicio 2: ¿Cuál es el promedio de puntuación de las películas agrupadas por clasificación de audiencia ('rating')? Interpreta los resultados, ¿hay algo que te llame la atención? Explica por qué. 

In [12]:
df.groupby("rating")["IMDB Score"].mean()

rating
66 min           NaN
74 min           NaN
84 min           NaN
G           6.500000
NC-17            NaN
NR               NaN
PG          6.800000
PG-13       6.386957
R           6.495745
TV-14       6.177660
TV-G        6.156250
TV-MA       6.097638
TV-PG       6.308772
TV-Y        6.780000
TV-Y7       6.240000
TV-Y7-FV         NaN
UR               NaN
Name: IMDB Score, dtype: float64

Cosas que nos llaman la antencion:
- Las puntuaciones son *muy similares*.
- Existen varias clasificaciones de la audiencia con *valores nulos*.
- Hay algunas clasificaciones que parecen *no ser correctas* y que deberian estar en el campo de duracion. 

#### Ejercicio 3: ¿Cuántos títulos hay por cada país? Ordena los resultados de mayor a menor

In [22]:
df_agrupado = df.groupby("country")["title"].count()
df_agrupado.sort_values()

country
, France, Algeria                            1
Spain, Germany, Denmark, United States       1
Spain, Italy, Argentina                      1
Spain, Mexico, France                        1
Spain, Portugal                              1
                                          ... 
South Korea                                199
Japan                                      245
United Kingdom                             419
India                                      972
United States                             2818
Name: title, Length: 748, dtype: int64

#### Ejercicio 4: ¿Cuál es la calificación promedio en IMDb de los títulos agrupados por tipo ('Movie' o 'TV Show')?

In [26]:
df.groupby("type")["IMDB Score"].mean()

type
Movie      6.210916
TV Show         NaN
Name: IMDB Score, dtype: float64

Los titulos de `TV Show` no parecen tener una calificaicon IMDB.

#### Ejercicio 5: ¿Cuántos títulos se añadieron a Netflix cada año? Para resolver este ejercicio tendrás que convertir la columna `date_added` a formato fecha usando el método `pd.to_datetime`. [Aquí](https://pandas.pydata.org/docs/reference/api/pandas.to_datetime.html) tienes la documentación, leeala con detenimiento antes de usar el método. 

In [33]:
# https://www.youtube.com/watch?v=9LbslWe9GXc  # para convertir a pd.to_datetime, ya que despues de leer la docu seguia con dudas
df["date_added"] = pd.to_datetime(df["date_added"], errors='coerce') # coarce es para que me continue si hay un error y ponga el valor como NAN
# https://sparkbyexamples.com/pandas/pandas-extract-year-from-datetime/ # para extraer el anio de las fechas
df['year_added'] = df['date_added'].dt.strftime('%Y')
df.groupby("year_added")["title"].count()

year_added
2008       2
2009       2
2010       1
2011      13
2012       3
2013      10
2014      23
2015      73
2016     418
2017    1164
2018    1625
2019    1999
2020    1878
2021    1498
Name: title, dtype: int64

#### Ejercicio 6: ¿Cuál es el país con más títulos lanzados en el año 2020? Para resolver este ejercicio después del gropuby os puede resultar interesante el método [`idxmax`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html)

In [40]:
df.groupby("release_year")["title"].count()
max_year = df.groupby("release_year")["title"].count().idxmax()

print(f"El año con mas lanzamientos fue el {max_year}")


El año con mas lanzamientos fue el 2018


#### Ejercicio 7:¿Cuál es el promedio de puntuación en IMDb de los títulos agrupados por país?

In [41]:
df.groupby("country")["IMDB Score"].mean()

country
, France, Algeria                                       NaN
, South Korea                                           NaN
Argentina                                              6.42
Argentina, Brazil, France, Poland, Germany, Denmark     NaN
Argentina, Chile                                        NaN
                                                       ... 
Venezuela                                               NaN
Venezuela, Colombia                                     NaN
Vietnam                                                 NaN
West Germany                                            NaN
Zimbabwe                                                NaN
Name: IMDB Score, Length: 748, dtype: float64

#### Ejercicio 8: ¿Cuántos títulos fueron lanzados en cada año agrupados por tipo ('Movie' o 'TV Show')?

In [ ]:
#df.groupby("release_year","type")["title"].count()

df2 = df.groupby(['release_year', 'type']).size()  # https://stackoverflow.com/questions/17679089/pandas-dataframe-groupby-two-columns-and-get-counts

df2

df3 = df.groupby(["release_year", "type"])["type"].count()

df3


release_year  type   
1925          TV Show      1
1942          Movie        2
1943          Movie        3
1944          Movie        3
1945          Movie        3
                        ... 
2019          TV Show    397
2020          Movie      517
              TV Show    436
2021          Movie      277
              TV Show    315
Name: type, Length: 119, dtype: int64

- En el ejercicio 8 he usado 2 metodos. Estaba primero teniendo problemas con el metodo de los apuntes, algo estaba haciendo mal.
- Buscando en google, he encontrazo size() que parece darme el mismo resultado (df2)
- Despues, nuevamente, he vuelto a revisar lo de poner dos indices en groupby y esta vez me ha funcionado (df3)

#### Ejercicio 9: ¿Cuál es el país con la mayor cantidad de títulos de clasificación 'TV-MA'?

In [70]:
df2 = df.groupby(["rating", "country"])["IMDB Score"].mean().reset_index(name='title_mean')

filtered_df2 = df2[df2["rating"] == "TV-MA"]

fila_maxima = filtered_df2.loc[filtered_df2["title_mean"].idxmax()]  ## esta parte me ha costado. Al final he usado loc para filtrar haya donde el title mean era maximo

print(fila_maxima)

rating                                         TV-MA
country       United Kingdom, Ukraine, United States
title_mean                                       8.4
Name: 900, dtype: object


#### Ejercicio 10. ¿Cuál es el promedio de duración de las películas agrupadas por país?

In [85]:
df10 = df.groupby(["country"])["Runtime"].mean().reset_index(name='Runtime_mean').round()

df10_filtered = df10[df10["Runtime_mean"] > 0]

df10_filtered.sort_values("Runtime_mean")

country  Runtime_mean
383               Pakistan, United States          15.0
280                       Ireland, Canada          30.0
690                 United States, Mexico          31.0
474                 Sweden, United States          40.0
686                  United States, Japan          58.0
..                                    ...           ...
728  United States, United Kingdom, India         130.0
56                Cambodia, United States         136.0
458                        Spain, Germany         139.0
554    United Kingdom, Hungary, Australia         140.0
376        Norway, Iceland, United States         144.0

[81 rows x 2 columns]

#### Ejercicio 11: ¿Cuál es el número total de títulos por país y clasificación de audiencia ('rating')?

In [87]:
df11 = df.groupby(["rating", "country"])["title"].sum().reset_index(name='title_sum')

df11

rating                       country  \
0       66 min                 United States   
1       74 min                 United States   
2       84 min                 United States   
3            G                        Canada   
4            G  Canada, Japan, United States   
...        ...                           ...   
1170  TV-Y7-FV                 United States   
1171  TV-Y7-FV          United States, Italy   
1172        UR                        France   
1173        UR        United Kingdom, France   
1174        UR                 United States   

                                 title_sum  
0     Louis C.K.: Live at the Comedy Store  
1                          Louis C.K. 2017  
2                    Louis C.K.: Hilarious  
3                                 Pup Star  
4                     Pup Star: World Tour  
...                                    ...  
1170                          Dear Dracula  
1171                          Leo the Lion  
1172                         Immoral Tales  
1173                              Sex Doll  
1174         You Don't Mess with the Zohan  

[1175 rows x 3 columns]

#### Ejercicio 12: ¿Cuántos títulos fueron lanzados antes del año 2000 agrupados por país?

In [ ]:
df12 = df.groupby(["release_year", "country"])["title"].count().reset_index(name='title_sum')
df12_filtered = df12[df12["release_year"] < 2000]
df12_filtered 
total12 = df12_filtered["title_sum"].sum()
total12

print(f"Antes del anio 2000, se lanzaron {total12} titulos")




Antes del anio 2020, se lanzaron 502 titulos


#### Ejercicio 13: ¿Cuál es la puntuación promedio en IMDb de los títulos lanzados en 2021 agrupados por país?

In [108]:
df13 = df.groupby(["country", "release_year"])["IMDB Score"].mean().reset_index(name='IMDB Score Media')

df13_filtered = df13[df13["release_year"] == 2021]

puntuacion = df13_filtered["IMDB Score Media"].mean()

puntuacion

print(f"Para el anio 2021, la puntuacio media fue de {puntuacion.round(1)}")

Para el anio 2021, la puntuacio media fue de 5.9


#### Ejercicio 14: ¿Cuántos títulos fueron dirigidos por cada director? Ordena los resultados de mayor a menor

In [113]:
df14 = df.groupby(["director"])["title"].count().reset_index(name='pelis_por_director')

df14.sort_values('pelis_por_director',ascending=False)

director  pelis_por_director
3392                          Rajiv Chilaka                  19
3443                 Raúl Campos, Jan Suter                  18
4046                            Suhas Kadav                  16
2598                           Marcus Raboy                  16
1790                              Jay Karas                  14
...                                     ...                 ...
2049                           Jos Humphrey                   1
2050                             Jose Gomez                   1
2051                      Jose Javier Reyes                   1
2053  Joseduardo Giordano, Sergio Goyri Jr.                   1
2264                         Khaled Youssef                   1

[4528 rows x 2 columns]

#### Ejercicio 15: ¿Cuál es el lenguaje más frecuente en los títulos lanzados en los últimos 5 años agrupado por país?

In [119]:
df14 = df.groupby(["Language","release_year"])["title"].count().reset_index(name='pelis_por_director')

filtered_df14 = df14[df14["release_year"] > 2016]

fila_maxima = filtered_df14.loc[filtered_df14["pelis_por_director"].idxmax()]  ## he repetido lo mismo que en el ejercicio 9

idioma_mas_hablado = fila_maxima["Language"]

print(f"El idioma mas hablado es `{idioma_mas_hablado}`")

El idioma mas hablado es `English`
